nutils/topology.py

```diff
@@ -1712,255 +1769,252 @@ def connectivity(self):
+    refine = numpy.unique(numpy.array(refine, dtype=int))
+    splits = numpy.searchsorted(refine, self._offsets, side='left')
+    indices_per_level = list(map(list, self._indices_per_level))+[[]]
+    fine = self.basetopo
+    for ilevel, (start, stop) in enumerate(zip(splits[:-1], splits[1:])):
+      coarse, fine = fine, fine.refined
+      coarse_indices = tuple(map(indices_per_level[ilevel].pop, reversed(refine[start:stop]-self._offsets[ilevel])))
+      coarse_transforms = map(coarse.transforms.__getitem__, coarse_indices)
+      coarse_references = map(coarse.references.__getitem__, coarse_indices)
+      fine_transforms = (trans+(ctrans,) for trans, ref in zip(coarse_transforms, coarse_references) for ctrans, cref in ref.children if cref)
+      indices_per_level[ilevel+1].extend(map(fine.transforms.index, fine_transforms))
+    if not indices_per_level[-1]:
+      indices_per_level.pop(-1)
+    return self.hierarchical([numpy.unique(numpy.array(i, dtype=int)) for i in indices_per_level])
+
+  def hierarchical(self, elements):
+    return self.basetopo.hierarchical(elements)
 
   @property
   def refined(self):
-    elements = [child for elem in self for child in elem.children]
-    return self.basetopo.hierarchical(elements, precise=True)
+    refined_indices_per_level = [[]]
+    fine = self.basetopo
+    for coarse_indices in self._indices_per_level:
+      coarse, fine = fine, fine.refined
+      coarse_transforms = map(coarse.transforms.__getitem__, coarse_indices)
+      coarse_references = map(coarse.references.__getitem__, coarse_indices)
+      fine_transforms = (trans+(ctrans,) for trans, ref in zip(coarse_transforms, coarse_references) for ctrans, cref in ref.children if cref)
+      refined_indices_per_level.append(numpy.unique(numpy.fromiter(map(fine.transforms.index, fine_transforms), dtype=int)))
+    return self.basetopo.hierarchical(refined_indices_per_level)
 
   @property
   @log.withcontext
   def boundary(self):
     'boundary elements'
 
     basebtopo = self.basetopo.boundary
-    edgepool = [edge for elem in self if transform.lookup(elem.transform, self.basetopo.border_transforms) for edge in elem.edges if edge is not None]
-    belems = []
-    for edge in edgepool: # superset of boundary elements
-      try:
-        iedge, tail = transform.lookup_item(edge.transform, basebtopo.edict)
-      except KeyError:
-        pass
-      else:
-        opptrans = basebtopo.elements[iedge].opposite + tail
-        belems.append(element.Element(edge.reference, edge.transform, opptrans))
-    return basebtopo.hierarchical(belems, precise=True)
+    bindices_per_level = []
+    blevel = None
+    for level, indices in zip(self.levels, self._indices_per_level):
+      blevel = basebtopo if blevel is None else blevel.refined
+      bindex = blevel.transforms.index
+      bindices = []
+      for trans, ref in zip(map(level.transforms.__getitem__, indices), map(level.references.__getitem__, indices)):
```

Comment:
Don't we loose performance here by looping over all elements instead of "edgepool"?  I imagine that concept can still be applied per level to reduce indices since we already have blevel.

Reply:
In `examples/adaptivity.py` creating the boundary is not so expensive. Not sure if this is worth the effort at this point.

Reply:
ok